tests/base.py

```diff
@@ -188,6 +188,27 @@ def assertEqualMem(self, a, b, ignore=None):
                 # in GMRS)
                 continue
 
+            # Special case for radios with "false" duplex=off functionality
+            # achieved through large offsets to get tx freq out-of-band
+            if k in ["offset", "duplex"] and \
+               hasattr(self.radio, "DUPLEX_OFF_VIA_OFFSET") and \
+               self.radio.DUPLEX_OFF_VIA_OFFSET:
+                matching_band = None
+                for i, limits in enumerate(self.radio.valid_bands):
+                    lo, hi = limits
+                    if lo <= a.freq <= hi:
+                        matching_band = i
+                        break
+                if matching_band is not None and \
+                   self.radio.DUPLEX_OFF_VIA_OFFSET[i] is not None:
+                    duplex, offset = self.radio.DUPLEX_OFF_VIA_OFFSET[i]
+                    if a.offset == offset and a.duplex == duplex and \
+                       b.duplex == "off":
+                        continue
+                    if b.offset == offset and b.duplex == duplex and \
+                       a.duplex == "off":
+                        continue
+
```

Comment:
First, major props for adding test coverage. *Nobody* helps me with test coverage :)

However, I think maybe we should put this in a different spot. You're checking for your flag, but that flag isn't on the base radio type, it's really an internal thing to just your driver. Like, if someone else just happened to implement that same flag in a way that means something very different, this might be triggered in a good or bad way. This is a pretty deep spot to embed that logic (even though I recognize it's not too dissimilar from some other things here).

So I think we should still do this check, but let's put it in the unit tests, where we have some more targeted tests for specific radios and situations, if that's okay. Here's an example of a radio-specific unit test file for some of the cloning routines in one driver:

https://github.com/kk7ds/chirp/blob/master/tests/unit/test_th9800.py

I think that sort of thing is best for here, that way you can write it targeted for just the ft4 and not add these assertions that could potentially affect everything. If so inclined, that would also give you a spot to add further test coverage for all the ft4 weirdness if you so chose :)

I'll be glad to move/rewrite this in that form for you if you like, happy to do that, but don't want to take and mangle your code if you'd rather.

Reply:
In this case, I had to modify `assertEqualMem` in order for most of the tests associated with the radio to pass, as many of the standard tests now trigger the special logic and fail the assertion because it's no longer a given that a `+` or `-` duplex will stay that way, it may instead become an `off`. Offsets as well would change in some circumstances. So the goal of this wasn't to be a unit test so much as it was to address the fact that this driver-specific logic affects some broad assertions - namely that for the FT-65R, out-of-bound offsets with a +/- duplex should be treated as equivalent to `duplex=off` when comparing memories.

If there's a good way to ensure that the ft4 driver tests have an overridden `assertEqualMem`, then I agree that this would belong somewhere driver-specific. Beyond that, I guess we could also branch on driver name for this special logic, so at least we don't run the risk of someone else using the same constant name for other purpose in a different driver (and then it would also be exceptionally clear that this quirk is driver-specific).

There's probably a clever way to inhibit those assertion fails in `assertEqualMem` and fill in equivalent driver-specific testing, but I wasn't coming up with anything when I tried.

Reply:
Ohh, I see, I read this too fast during dinner. You're bailing in this scenario and not running stuff below, I see, I was thinking this was just checking those limits, but now I see what's going on.

The problem with that is that you're basically eliminating all the test assertions for radios with that property, which is, uh, cheating :)

Let me think on this a bit tomorrow and see what I can come up with. In general, we should not be modifying the tests, but making the driver do the right thing of course.

Reply:
Okay, so, with this test change reverted, I see failing one test:
```
FAILED tests/test_drivers.py::TestCaseEdges_Yaesu_FT-65R::test_get_set_specials - AssertionError: {'fre[67 chars]x': 'off'...
```

And it looks like a legit fail to me. The test tried to set a `duplex=off` memory, but got back a `duplex=+` memory. That's because while you converted `duplex=off` to "big offset" in `set_memory()`, your `get_memory()` didn't turn that back into `duplex=off` on the way back out. So that's a bug in your driver, as get/set should be identical in this case.

We can make `assertEqualMem()` ignore the offset discrepancy if `duplex=off`, but that's not the actual problem here that I see.